### ISSUE TITLE
[phase-2] DSpark C4 — draft-head training harness on the OPD substrate; first artifact: DSv4-Flash head

### ISSUE THREAD

**Issue by cklxx** [cklxx/agent-infer#127]
Umbrella: #123. DSv4-Flash has no public DSpark head — DeepSpec ships the training recipe, and ARLE's OPD substrate (teacher surface on `infer-api` + `crates/autograd`) is exactly the differentiated loop for training one: strong teacher inference + tight student-scoring latency, train-infer-unified.

## Work

- Training harness for the DSpark head triple on the OPD substrate: parallel backbone (`draft_hidden~1024`, ~5 layers, `block_size~7`, target hidden taps from a few target layers) + Markov/RNN sequential head + confidence head.
- Teacher = the serving DSv4-Flash engine via the existing OPD-teacher surface — no PyTorch, no Python on the hot path.
- Recipe fidelity per the C2 component map; deviations are deliberate commits, not improvisation.
- First artifact: a DSv4-Flash draft head consumable by the C3 lane (and exportable for C6-style 4-bit conversion).

## Exit condition (license-or-kill)

- Trained head reaches accept-length ≥ the H20 break-even (~57%) on the SLO workload — the number that decides whether C5 can ever pay.
- On a miss: **decode the failing cases before any structural kill** (case-as-fact rule, AGENTS.md §0) — draft-quality attribution at the token level, audit the eval harness for artifacts, then verdict.
- Wins/errors entry with the acceptance curve either way.

**Comment by cklxx** [comment]
Premise correction from the C2 scout (#125): **DSv4-Flash has a public official head** — `deepseek-ai/DeepSeek-V4-Flash-DSpark` on HF. Per adopt-official-first, C5 (#128) consumes that head directly. This issue re-scopes to: (a) training harness for the no-public-head targets — Qwen3.6/Qwen3.5-MoE (feeds C6 #129), Qwen3.5 hybrid (C7 #130), GLM-5.2; (b) fine-tune fallback if the adopted V4-Flash head's acceptance underperforms the paper claim on our serving distribution. Vendoring the DeepSpec Python trainer happens at this issue's start, pinned from afdfa7c9382a.

**Comment by cklxx** [comment]
**The premise is false: this head does not need training. It is already published, MIT-licensed, in our loader's exact schema.**

Verified directly today (not inferred): [`deepseek-ai/DeepSeek-V4-Flash-DSpark`](https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-DSpark) — 167 GB / 48 shards, MIT. Its [config.json](https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-DSpark/raw/main/config.json) carries **all four keys `merge_dspark_metadata` requires, under exactly our names**:

```
dspark_block_size: 5
dspark_target_layer_ids: [40, 41, 42]
dspark_markov_rank: 256
dspark_noise_token_id: 128799
hidden_size: 4096   num_hidden_layers: 43   vocab_size: 129280
expert_dtype: "fp4"  quantization_config: fp8 e4m3 / ue8m0 / block [128,128]
```

The README states it plainly: *"not a new model. It is the same checkpoint with an additional speculative decoding module attached."*

So the work is a **weight-import + conversion job, not a training job.** Recommend rescoping this issue accordingly (or closing it in favour of #128, whose precondition "C4 head at ≥ break-even acceptance" becomes nearly free).

### Why the training framing looked necessary, and why it wasn't

`DeepSpec` (MIT, DeepSeek's own full-stack trainer) publishes configs for **Qwen3-4B/8B/14B and Gemma-4-12B only — no V4 config**. That is what our docs recorded as "DSv4 not open-sourced", and the imprecision matters: the *training code* for V4 is unpublished, the *weights* are not. Training a V4 head ourselves would mean writing the DSv4 target adapter (MLA + mHC taps + fp4/fp8 experts) inside DeepSpec, regenerating ~1.3M responses through a 284B-A13B MoE, and storing the hidden-state cache — DeepSpec's own default cache is ~38 TB for the *Qwen3-4B* setting, against 617 GB free on our pod at 0.2 GB/s. That cost buys a checkpoint that already exists.

A from-scratch Rust/autograd trainer is worse still: our autograd crate is host-f32 AdamW with no FP8 GEMM, no expert parallelism and no MoE backward, so "train a draft head" decomposes into "build a distributed MoE trainer". The "no PyTorch/no Python" rule protects the **inference hot path**; a one-time offline job emitting safetensors violates nothing. Recommend killing that direction explicitly.

### Also settled: the "draft_hidden ~1024" spec in this issue is wrong

It came from the paper sketch in `docs/architecture-dsv4.md` §7, which is labelled source-survey. Real artifacts are **trunk-width with few layers**: the V4 head is hidden 4096 (= trunk), and `z-lab/Qwen3.6-27B-DFlash` is hidden **5120** / 5 layers (= its trunk) — which is why our `ensure!(cfg.hidden_size == trunk_hidden)` never tripped on it. No published design runs the backbone narrower than the trunk; the shared frozen embed/LM-head is the structural reason. Our loader's shape was right all along, and the doc's 1024 should be corrected.

### What we would still have to build (the real residual)

1. **Confidence *scheduler*** — the head's confidence *weights* ship; the per-request variable-length verify budget + calibration is inference policy, i.e. #124's territory.
2. **fp4-expert / fp8-block-scale handling for the draft's expert tensors** — same scheme as the trunk, but a conversion/loader pass to confirm.
3. **Ragged verify plumbing** for per-request draft lengths.

### Two facts I could NOT establish — check before writing loader logic

- **The stage count.** My earlier note recorded one stage (`mtp.0`); a survey today read three (`mtp.0/1/2`) off a third-party mirror. I did not list tensor names in the official shards myself. `count_dspark_stages` counts at load time, so don't hardcode either number.
- **The repo tree shows no `model.safetensors.index.json`**, so `count_dspark_stages` must fall back to reading shard headers. Worth confirming that path works before assuming the import is clean.

Cheapest next step: fetch the ~10 GB draft-delta shards (46-48 of 48), list their tensor names, and attempt a load — that resolves both unknowns and either validates or kills the import in an afternoon.

https://claude.ai/code/session_016ZQ8zX4gdJ7jChc2GAU16M

**Comment by cklxx** [comment]
Done: DSv4-Flash head adopted directly (public: deepseek-ai/DeepSeek-V4-Flash-DSpark, 4705 tensors); training harness shipped as fine-tune fallback (crates/train/src/dspark_train.rs:822). CHANGELOG REJECT 'train a DSv4 draft head'.
